configuration/postgres/postgres.go

```diff
@@ -0,0 +1,417 @@
+	connectionStringKey          = "connectionString"
+	pgNotifyChannelKey           = "pgNotifyChannel"
+	listenTemplate               = "listen %s"
+	unlistenTemplate             = "unlisten %s"
+	payloadDataKey               = "data"
+	ErrorMissingTableName        = "missing postgreSQL configuration table name"
+	ErrorMissingTable            = "postgreSQL configuration table - '%s' does not exist"
+	InfoStartInit                = "initializing postgreSQL configuration store"
+	ErrorMissingConnectionString = "missing postgreSQL connection string"
+	ErrorAlreadyInitialized      = "postgreSQL configuration store already initialized"
+	ErrorMissingMaxTimeout       = "missing PostgreSQL maxTimeout setting in configuration"
+	QueryTableExists             = "SELECT EXISTS (SELECT FROM pg_tables where tablename = $1)"
+	ErrorTooLongFieldLength      = "field name is too long"
+	maxIdentifierLength          = 64 // https://www.postgresql.org/docs/current/limits.html
+)
+
+var (
+	allowedChars           = regexp.MustCompile(`^[a-zA-Z0-9./_]*$`)
+	defaultMaxConnIdleTime = time.Minute * 30
+)
+
+func NewPostgresConfigurationStore(logger logger.Logger) configuration.Store {
+	logger.Debug("Instantiating PostgreSQL configuration store")
+	return &ConfigurationStore{
+		logger:               logger,
+		subscribeStopChanMap: make(map[string]interface{}),
+		configLock:           sync.Mutex{},
+	}
+}
+
+func (p *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	p.logger.Debug(InfoStartInit)
+	if p.client != nil {
+		return fmt.Errorf(ErrorAlreadyInitialized)
+	}
+	if m, err := parseMetadata(metadata); err != nil {
+		p.logger.Error(err)
+		return err
+	} else {
+		p.metadata = m
+	}
+	p.ActiveSubscriptions = make(map[string]*subscription)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultMaxConnIdleTime)
+	defer cancel()
+	// ctx := context.Background()
+	client, err := Connect(ctx, p.metadata.connectionString)
+	if err != nil {
+		return fmt.Errorf("error connecting to configuration store: '%s'", err)
+	}
+	p.client = client
+	pingErr := p.client.Ping(ctx)
+	if pingErr != nil {
+		return fmt.Errorf("unable to connect to configuration store: '%s'", pingErr)
+	}
+	// check if table exists
+	exists := false
+	err = p.client.QueryRow(ctx, QueryTableExists, p.metadata.configTable).Scan(&exists)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return fmt.Errorf(ErrorMissingTable, p.metadata.configTable)
+		}
+		return fmt.Errorf("error in checking if configtable exists - '%v'", p.metadata.configTable)
+	}
+	return nil
+}
+
+func (p *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequest) (*configuration.GetResponse, error) {
+	if err := validateInput(req.Keys); err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	query, params, err := buildQuery(req, p.metadata.configTable)
+	if err != nil {
+		p.logger.Error(err)
+		return nil, fmt.Errorf("error in configuration store query: '%s' ", err)
+	}
```

Comment:
With `fmt.Errorf`, when you pass errors, it's best to wrap them using `%w` which allows the error to be un-wrapped. See: https://gosamples.dev/wrap-unwrap-errors/

Reply:
done